syntax = "proto3";

package messenger;

import "google/protobuf/timestamp.proto";
import "common.proto";

// messages the server sends to the client
message ServerMessage {
  message UserLogInResponse { bool is_ok = 1; }

  message ThreadMessagePack {
    message ThreadSendMessageRequest {
      uint64 id = 1;
      uint64 thread_id = 2;
      string body = 3;
      google.protobuf.Timestamp time = 4;
    }
    repeated ThreadSendMessageRequest messages = 1;
  }

  message ThreadPack {
    message ThreadData {
      uint64 id = 1;
      string head = 2;
      string body = 3;
    }
    repeated ThreadData threads = 1;
  }

  oneof inner {
    UserCredentials new_account_data = 1;
    UserLogInResponse user_log_in_response = 2;
    ThreadMessagePack new_messages_appeared = 3;

    // server sends all the threads to the client on any change (TODO)
    ThreadPack all_the_threads = 4;
  }
}